### PR TITLE
Highnodeutilization strategy

### LINKED ISSUE #196: Add HighNodeUtilization strategy
I want to keep my cluster as compact as possible. In order to do that, I switched kubernetes-scheduler policy from LeastRequestedPriority to MostRequestedPriority. It would be greate to have corresponding HighNodeUtilization strategy for descheduler.

### PR DESCRIPTION
**Pull request by hanumanthan**
Adds HighNodeUtilization strategy

Closes #196

### CODE REVIEW

**Comment by ingvagabund** [comment]
To make the review easier once the PR is not WIP, can you create a separate commit just for moving the code (without renaming it, without changing anything else)?

**Comment by hanumanthan** [comment]
> To make the review easier once the PR is not WIP, can you create a separate commit just for moving the code (without renaming it, without changing anything else)?

Sure, I will seperate into two commits - one for moving common functions and other for high utilisation. I have created this PR and marked as WIP just to keep track of any changes that happens in the upstream repo

**Comment by hanumanthan** [comment]
/assign @damemi @ingvagabund 

Can you please review this PR

Is `e2e test` applicable ? I tried to run and few of the existing tests are failing. Not sure if its up to date and this strategy should be added to it

**Comment by hanumanthan** [comment]
Hi @damemi @ingvagabund Could you please review this ? Thanks

File: pkg/descheduler/strategies/nodeutilization.go [hunk 1]
```diff
@@ -149,38 +149,40 @@ func resourceUsagePercentages(nodeUsage NodeUsage) map[v1.ResourceName]float64 {
 	return resourceUsagePercentage
 }
 
-// classifyNodes classifies the nodes into low-utilization or high-utilization nodes. If a node lies between
-// low and high thresholds, it is simply ignored.
+// classifyNodes classifies the nodes into target node or probable node based on the threshold filter functions.
+// Nodes not matching both filters are simply ignored
```
Comment (ingvagabund) [inline pkg/descheduler/strategies/nodeutilization.go:153]: What is meaning of target node and probable node in context of node's utilization? `classifyNodes` classifies nodes based on utilization thresholds so it's still valid to use low/high keyword. How is this change related to implementation of high/low node utilization strategies?

File: README.md [hunk 2]
```diff
@@ -232,6 +233,58 @@ This parameter can be configured to activate the strategy only when the number o
 are above the configured value. This could be helpful in large clusters where a few nodes could go
 under utilized frequently or for a short period of time. By default, `numberOfNodes` is set to zero.
 
+### HighNodeUtilization
+
+This strategy finds nodes that are under utilized and evicts pods in the hope that these pods will be scheduled compactly into fewer nodes. This strategy can be used in tandem with the 
+scheduler strategy `MostRequestedPriority`. The parameters of this strategy are configured under `nodeResourceUtilizationThresholds`.
```
Comment (damemi) [inline README.md:239]: Ideally this strategy *must* be used with `MostRequestedPriority` (actually called `NodeResourcesMostAllocated` in the [plugin framework](https://kubernetes.io/docs/reference/scheduling/config/#scheduling-plugins-1)). Using `HighNodeUtilization` without the matching scheduler plugin will cause the evicted pods to be rescheduled back onto the lowest-utilized nodes

File: README.md [hunk 3]
```diff
@@ -106,7 +107,7 @@ See the [user guide](docs/user-guide.md) in the `/docs` directory.
 ## Policy and Strategies
 
 Descheduler's policy is configurable and includes strategies that can be enabled or disabled.
-Eight strategies `RemoveDuplicates`, `LowNodeUtilization`, `RemovePodsViolatingInterPodAntiAffinity`,
+Nine strategies `RemoveDuplicates`, `LowNodeUtilization`, `HighNodeUtilization`, `RemovePodsViolatingInterPodAntiAffinity`,
```
Comment (damemi) [inline README.md:110]: We could probably update this list to be more readable (maybe bulleted/numbered?)

File: pkg/descheduler/strategies/nodeutilization.go (hunk 1 above)
Comment (hanumanthan) [reply pkg/descheduler/strategies/nodeutilization.go:153]: low and high node representation matched the `threshold` and `targetthreshold` in LowNodeUtilization. For HighNodeUtilization, as there is only one threshold - low nodes, should we consider the rest as high nodes ?
I tried to name them as target and probable to convey if a node is considered for pod eviction or its a node where the evicted pods could end

Comment (ingvagabund) [reply pkg/descheduler/strategies/nodeutilization.go:153]: > as there is only one threshold

for now though in the future it might be configurable as well

Given LNU still invokes `classifyNodes`, the semantic needs to stay the same. It's also up to a caller of `classifyNodes` to decide which arguments to pass. I still don't see a benefit of renaming the parameters or changing the order in which `classifyNodes` returns the values. Please, keep it as it is.

File: pkg/descheduler/strategies/nodeutilization.go [hunk 4]
```diff
@@ -149,38 +149,40 @@ func resourceUsagePercentages(nodeUsage NodeUsage) map[v1.ResourceName]float64 {
...
 // TODO: @ravig Break this function into smaller functions.
 func evictPodsFromTargetNodes(
 	ctx context.Context,
-	targetNodes, lowNodes []NodeUsage,
+	targetNodes, probableNodes []NodeUsage,
 	podEvictor *evictions.PodEvictor,
 	podFilter func(pod *v1.Pod) bool,
 	resourceNames []v1.ResourceName,
+	strategy string,
+	continueEvictionCond func(nodeUsage NodeUsage, totalAvailableUsage map[v1.ResourceName]*resource.Quantity) bool,
```
Comment (ingvagabund) [inline pkg/descheduler/strategies/nodeutilization.go:185]: Can you define:
```
type continueEvictionCond func(nodeUsage NodeUsage, totalAvailableUsage map[v1.ResourceName]*resource.Quantity) bool
```
and pass the type instead?

File: pkg/descheduler/strategies/nodeutilization.go [hunk 5]
```diff
@@ -299,8 +290,8 @@ func evictPods(
 				}
 
 				klog.V(3).InfoS("Updated node usage", keysAndValues...)
-				// check if node utilization drops below target threshold or any required capacity (cpu, memory, pods) is moved
-				if !continueCond(nodeUsage, totalAvailableUsage) {
+				// check if the eviction conditions still holds true
```
Comment (ingvagabund) [inline pkg/descheduler/strategies/nodeutilization.go:293]: // check if pods can be still evicted

File: pkg/descheduler/strategies/highnodeutilization.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (ingvagabund) [inline pkg/descheduler/strategies/highnodeutilization.go:2]: 2021

File: pkg/descheduler/strategies/highnodeutilization.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+	"k8s.io/apimachinery/pkg/api/resource"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/descheduler/pkg/api"
+	"sigs.k8s.io/descheduler/pkg/descheduler/evictions"
+	nodeutil "sigs.k8s.io/descheduler/pkg/descheduler/node"
+	"sigs.k8s.io/descheduler/pkg/utils"
+)
+
+// HighNodeUtilization evicts pods from under utilised nodes so that scheduler can schedule according to its strategy.
```
Comment (ingvagabund) [inline pkg/descheduler/strategies/highnodeutilization.go:32]: utilised -> utilized

File: pkg/descheduler/strategies/lownodeutilization_test.go [hunk 8]
```diff
@@ -752,95 +745,7 @@ func TestValidateStrategyConfig(t *testing.T) {
 	}
 }
 
-func TestValidateThresholds(t *testing.T) {
```
Comment (ingvagabund) [inline pkg/descheduler/strategies/lownodeutilization_test.go:755]: Please carry this change in the first commit.

File: pkg/descheduler/strategies/lownodeutilization_test.go [hunk 9]
```diff
@@ -598,7 +591,7 @@ func TestLowNodeUtilization(t *testing.T) {
 	}
 }
 
-func TestValidateStrategyConfig(t *testing.T) {
+func TestValidateLowNodeUtilizationStrategyConfig(t *testing.T) {
```
Comment (ingvagabund) [inline pkg/descheduler/strategies/lownodeutilization_test.go:594]: Please carry this change in the second commit.

File: pkg/descheduler/strategies/lownodeutilization_test.go [hunk 10]
```diff
@@ -970,41 +875,3 @@ func TestWithTaints(t *testing.T) {
 		})
 	}
 }
-
-func TestResourceUsagePercentages(t *testing.T) {
```
Comment (ingvagabund) [inline pkg/descheduler/strategies/lownodeutilization_test.go:974]: Please carry this change in the first commit.

File: pkg/descheduler/strategies/lownodeutilization_test.go [hunk 11]
```diff
@@ -498,7 +491,7 @@ func TestLowNodeUtilization(t *testing.T) {
 						*test.BuildTestPod("p1", 0, 0, n1NodeName, func(pod *v1.Pod) {
 							// A pod with extended resource.
 							test.SetRSOwnerRef(pod)
-							test.SetPodExtendedResourceRequest(pod, extendedResource, 1)
+							test.SetPodExtendedResourceRequest(pod, extendedResource, 5)
```
Comment (ingvagabund) [inline pkg/descheduler/strategies/lownodeutilization_test.go:494]: Why 5?

File: pkg/descheduler/strategies/highnodeutilization.go [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+		return
+	}
+	thresholdPriority, err := utils.GetPriorityFromStrategyParams(ctx, client, strategy.Params)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get threshold priority from strategy's params")
+		return
+	}
+
+	thresholds := strategy.Params.NodeResourceUtilizationThresholds.Thresholds
+	targetThresholds := strategy.Params.NodeResourceUtilizationThresholds.TargetThresholds
```
Comment (ingvagabund) [inline pkg/descheduler/strategies/highnodeutilization.go:46]: - `thresholds` is the low threshold
- `targetThresholds` is the high threshold

As discussed in https://github.com/kubernetes-sigs/descheduler/issues/196#issuecomment-814816526, please pin targetThresholds of all resources to 100%. Also, `validateNodeUtilizationParams` needs to make sure `strategy.Params.NodeResourceUtilizationThresholds.TargetThresholds` is never set until we allow it.

File: pkg/descheduler/strategies/highnodeutilization.go [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+
+	setDefaultForThresholds(thresholds, targetThresholds)
+	resourceNames := getResourceNames(targetThresholds)
+
+	targetNodes, probableNodes := classifyNodes(
+		getNodeUsage(ctx, client, nodes, targetThresholds, thresholds, resourceNames),
+		func(node *v1.Node, usage NodeUsage) bool {
+			return isNodeWithLowUtilization(usage)
+		},
+		func(node *v1.Node, usage NodeUsage) bool {
```
Comment (ingvagabund) [inline pkg/descheduler/strategies/highnodeutilization.go:61]: Both `lowThresholdFilter` and `highThresholdFilter` have to have disjoint intervals.
```
func(node *v1.Node, usage NodeUsage) bool {
	if nodeutil.IsNodeUnschedulable(node) {
		klog.V(2).InfoS("Node is unschedulable", "node", klog.KObj(node))
		return false
	}
	return true
})
```
includes
```
func(node *v1.Node, usage NodeUsage) bool {
	return isNodeWithLowUtilization(usage)
},
```
as well.

**Review by ingvagabund — CHANGES_REQUESTED** [review]

**Comment by hanumanthan** [comment]
Addressed the review comments. @ingvagabund please have a look

**Comment by damemi** [comment]
/ok-to-test
@hanumanthan could you rebase? Sorry for the delay

File: pkg/descheduler/strategies/nodeutilization/lownodeutilization.go [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+		targetThresholds[v1.ResourcePods] = MaxResourcePercentage
+	}
+	if _, ok := thresholds[v1.ResourceCPU]; !ok {
+		thresholds[v1.ResourceCPU] = MaxResourcePercentage
+		targetThresholds[v1.ResourceCPU] = MaxResourcePercentage
+	}
+	if _, ok := thresholds[v1.ResourceMemory]; !ok {
+		thresholds[v1.ResourceMemory] = MaxResourcePercentage
+		targetThresholds[v1.ResourceMemory] = MaxResourcePercentage
+	}
```
[the remark this answers is not in the document]
Comment (hanumanthan) [reply pkg/descheduler/strategies/nodeutilization/lownodeutilization.go:65]: This code block sets default to resources that are not specified in the config. While `setDefaultForThresholds` sets default for all targetThreshold. I couldn't think of a better way to acheive that, so left things as it is

**Comment by hanumanthan** [comment]
> Would like to merge this soon (this week) to include it in the 1.21 release.

I am interested to participate in the release process to learn how they are done. Any place to start watching and participating ?

**Comment by damemi** [comment]
@hanumanthan we track each release in a pinned issue, 1.21 is https://github.com/kubernetes-sigs/descheduler/issues/491 and docs on the process are at https://github.com/kubernetes-sigs/descheduler/blob/master/docs/release-guide.md. It is a semi-manual process of pushing a tag to the Git repo, updating the published images on gcr.io, and publishing the release on Github

**Comment by hanumanthan** [comment]
Gentle reminder for a review. Thanks

File: pkg/descheduler/strategies/nodeutilization/nodeutilization_test.go [hunk 15]
```diff
@@ -25,6 +25,12 @@ import (
 	"testing"
 )
 
+var (
```
Comment (ingvagabund) [inline pkg/descheduler/strategies/nodeutilization/nodeutilization_test.go:28]: Please add this change to the first commit.

File: README.md [hunk 16]
```diff
@@ -237,6 +246,58 @@ This parameter can be configured to activate the strategy only when the number o
 are above the configured value. This could be helpful in large clusters where a few nodes could go
 under utilized frequently or for a short period of time. By default, `numberOfNodes` is set to zero.
 
+### HighNodeUtilization
+
+This strategy finds nodes that are under utilized and evicts pods in the hope that these pods will be scheduled compactly into fewer nodes. This strategy **must** be used with the 
+scheduler strategy `MostRequestedPriority`. The parameters of this strategy are configured under `nodeResourceUtilizationThresholds`.
+
+The under utilization of nodes is determined by a configurable threshold `thresholds`. The threshold
+`thresholds` can be configured for cpu, memory, number of pods, and extended resources in terms of percentage (the percentage is
```
Comment (ingvagabund) [inline README.md:258]: No need to put `the percentage is` sentence into parenthesis. The meaning is kept even without it. I.e.
`The percentage is ... pod capacity set for that node.`

File: README.md [hunk 17]
```diff
@@ -237,6 +246,58 @@ This parameter can be configured to activate the strategy only when the number o
...
+calculated as the current resources requested on the node vs [total allocatable](https://kubernetes.io/docs/concepts/architecture/nodes/#capacity).
+For pods, this means the number of pods on the node as a fraction of the pod capacity set for that node).
+
+If a node's usage is below threshold for all (cpu, memory, number of pods and extended resources), the node is considered underutilized.
+Currently, pods request resource requirements are considered for computing node resource utilization. Any node above `thresholds` is considered appropriately utilized and is not considered for eviction. 
+
+`thresholds` could be tuned as per your cluster requirements. Note that this
```
Comment (ingvagabund) [inline README.md:265]: The `threshold` param ...

File: README.md [hunk 18]
```diff
@@ -237,6 +246,58 @@ This parameter can be configured to activate the strategy only when the number o
...
+strategy evicts pods from `underutilized nodes` (those with usage below `thresholds`) so that they can be recreated in appropriately utilized nodes, it will abort if any number of `underutilized nodes` or `appropriately utilized nodes` is zero.
```
Comment (ingvagabund) [inline README.md:266]: `, it` -> `. The strategy`

File: README.md [hunk 19]
```diff
@@ -237,6 +246,58 @@ This parameter can be configured to activate the strategy only when the number o
...
+       nodeResourceUtilizationThresholds:
+         thresholds:
+           "cpu" : 20
+           "memory": 20
+           "pods": 20
+```
+
+Policy should pass the following validation checks:
+* Three basic native types of resources are supported: `cpu`, `memory` and `pods`. If any of these resource types is not specified, all its thresholds default to 100%.
+* Extended resources are supported. For example, resource type `nvidia.com/gpu` is specified for GPU node utilization. Extended resources are optional, and will not be used to compute node's usage if it's not specified in `targetThresholds` explicitly.
```
Comment (ingvagabund) [inline README.md:295]: `targetThresholds` -> `thresholds`

File: README.md [hunk 20]
```diff
@@ -237,6 +246,58 @@ This parameter can be configured to activate the strategy only when the number o
...
+* `thresholds` can not be nil.
+* The valid range of the resource's percentage value is \[0, 100\]
+
+There is another parameter associated with the `HighNodeUtilization` strategy, called `numberOfNodes`.
+This parameter can be configured to activate the strategy only when the number of under utilized nodes
+are above the configured value. This could be helpful in large clusters where a few nodes could go
```
Comment (ingvagabund) [inline README.md:301]: `are` -> `is`

File: pkg/descheduler/strategies/nodeutilization/highnodeutilization.go [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+		thresholds[v1.ResourcePods] = MaxResourcePercentage
+	}
+	if _, ok := thresholds[v1.ResourceCPU]; !ok {
+		thresholds[v1.ResourceCPU] = MaxResourcePercentage
+	}
+	if _, ok := thresholds[v1.ResourceMemory]; !ok {
+		thresholds[v1.ResourceMemory] = MaxResourcePercentage
+	}
+
+	//Default targetThreshold resource values to 100
```
Comment (ingvagabund) [inline pkg/descheduler/strategies/nodeutilization/highnodeutilization.go:147]: `//Default` -> `// Default`

File: pkg/descheduler/strategies/nodeutilization/highnodeutilization.go [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+	if err := validateHighUtilizationStrategyConfig(thresholds, targetThresholds); err != nil {
+		klog.ErrorS(err, "HighNodeUtilization config is not valid")
+		return
+	}
+	targetThresholds = make(api.ResourceThresholds)
+
+	setDefaultForThresholds(thresholds, targetThresholds)
+	resourceNames := getResourceNames(targetThresholds)
+
+	targetNodes, highNodes := classifyNodes(
```
Comment (ingvagabund) [inline pkg/descheduler/strategies/nodeutilization/highnodeutilization.go:56]: `s/targetNodes/lowNodes`

File: pkg/descheduler/strategies/nodeutilization/highnodeutilization.go [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+	continueEvictionCond := func(nodeUsage NodeUsage, totalAvailableUsage map[v1.ResourceName]*resource.Quantity) bool {
+		for name := range totalAvailableUsage {
+			if totalAvailableUsage[name].CmpInt64(0) < 1 {
+				return false
+			}
+		}
+
+		return true
+	}
+	evictPodsFromTargetNodes(
```
Comment (ingvagabund) [inline pkg/descheduler/strategies/nodeutilization/highnodeutilization.go:113]: `evictPodsFromTargetNodes` might be renamed to `evictPodsFromSourceNodes` (as part of the second commit)

File: pkg/descheduler/strategies/nodeutilization/lownodeutilization_test.go [hunk 24]
```diff
@@ -746,7 +746,7 @@ func TestValidateLowNodeUtilizationStrategyConfig(t *testing.T) {
 	}
 }
 
-func TestWithTaints(t *testing.T) {
+func TestLowNodeUtilizationWithTaints(t *testing.T) {
```
Comment (ingvagabund) [inline pkg/descheduler/strategies/nodeutilization/lownodeutilization_test.go:975]: Please move this change into the second commit.

**Review by ingvagabund — CHANGES_REQUESTED** [review]

File: test/e2e/e2e_test.go [hunk 25]
```diff
@@ -21,6 +21,7 @@ import (
 	"fmt"
 	"math"
 	"os"
+	"sigs.k8s.io/descheduler/pkg/descheduler/strategies/nodeutilization"
```
Comment (ingvagabund) [inline test/e2e/e2e_test.go:24]: Please put this alongside `sigs.k8s.io/descheduler/pkg` imports below.

File: pkg/descheduler/strategies/nodeutilization/highnodeutilization.go (hunk 22 above)
Comment (hanumanthan) [reply pkg/descheduler/strategies/nodeutilization/highnodeutilization.go]: Changed `targetNodes` to `sourceNodes` to match https://github.com/kubernetes-sigs/descheduler/pull/550#discussion_r643097917 and https://github.com/kubernetes-sigs/descheduler/pull/550#discussion_r622794415

**Comment by damemi** [comment]
/approve
@ingvagabund anything else you'd like to change?

File: pkg/descheduler/strategies/nodeutilization/highnodeutilization.go [hunk 26]
```diff
@@ -0,0 +1,157 @@
...
+	if len(sourceNodes) == len(nodes) {
+		klog.V(1).InfoS("All nodes are underutilized, nothing to do here")
+		return
+	}
+	if len(highNodes) == 0 {
+		klog.V(1).InfoS("No node is available to schedule the pods, nothing to do here")
+		return
+	}
+
+	evictable := podEvictor.Evictable(evictions.WithPriorityThreshold(thresholdPriority))
```
Comment (ingvagabund) [inline pkg/descheduler/strategies/nodeutilization/highnodeutilization.go:101]: Nit: missing WithNodeFit though it can be added in another PR:
```
evictable := podEvictor.Evictable(evictions.WithPriorityThreshold(thresholdPriority), evictions.WithNodeFit(nodeFit))
```

**Review by ingvagabund — APPROVED** [review]
Just a nit

**Comment by ingvagabund** [comment]
@hanumanthan thanks for bearing with us!!!
